---
title: "Final visualizations"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---


```{r, message=FALSE}
rm(list=ls())
library(Seurat)
library(ggplot2)

library(readxl)
library(writexl)
library(knitr)
library(data.table)
library(cowplot)
library(magrittr)
library(dplyr)

library(homologene)

library(pheatmap)

library(grid)
library(png)

data_storage_folder = "../../../Results/Jasso_Kinchen_combined"
result_folder = data_storage_folder

library(future)
options("future.globals.maxSize"=2**33)
#plan("multisession", workers = 8)
plan("sequential")

options("save.defaults" = list(compress=TRUE))
```



```{r}
load(file.path(data_storage_folder, "GSE114374_and_GSE172261_mouse_healthy_SCT.RData"))
cluster_assignment = read.table("../../../Data/cluster_assignments/stroma_in_vivo/Clusters_Kinchen_Jasso_2023-02-28.txt", sep="\t",header=T, stringsAsFactors = F)
scData.combined$cluster_label_orig = scData.combined$cluster_label
#options(width = 180)
#print(cluster_assignment)
rownames(cluster_assignment) = as.character(cluster_assignment$ClusterID)
scData.combined = AddMetaData(scData.combined, cluster_assignment[as.character(scData.combined$seurat_clusters),"Label"], col.name = "cluster_label")
scData.combined = AddMetaData(scData.combined, cluster_assignment[as.character(scData.combined$seurat_clusters),"Type"], col.name = "cluster_type")

scData.combined$cluster_label = factor(scData.combined$cluster_label, levels=sort(unique(scData.combined$cluster_label)))


scData.combined = SetIdent(scData.combined, value="cluster_label")
DefaultAssay(scData.combined) <- "RNA"

```


```{r, fig.width=10, fig.height=6}
p = DimPlot(scData.combined, reduction = "umap", label = TRUE, group.by = "cluster_label_orig") + NoLegend() + ggtitle("Jasso et al. original cluster labels")
print(p)

p = DimPlot(scData.combined, reduction = "umap", label = TRUE) + NoLegend() + ggtitle("Kinchen + Jasso et al. simplified clusters")
print(p)
```

# Fig. 2

## 2a

```{r}
p = DimPlot(scData.combined, label=T, repel=F) + NoLegend() + ggtitle("Kinchen et al. and Jasso et al. stroma integration")
print(p)

pdf(file.path(result_folder, "Fig_2a.pdf"), width = 6, height = 6)
print(p)
dev.off()

```

## Fig 2b

```{r, fig.width=7, fig.height=5}
p = DotPlot(scData.combined, features = c("Pecam1","Col1a2","Pdgfra","Acta2","S100b","Elavl4","Gli1"), assay = "RNA", cols="RdYlBu") + theme(axis.text.x = element_text(angle=45, hjust=1))
print(p)

pdf(file.path(result_folder, "Fig_2b.pdf"), width = 7, height = 5)
print(p)
dev.off()

```

## Fig 2e

```{r, fig.width=10, fig.height=7}
sel_markers =  c("Foxl1","Postn","Wnt5a","Cd34","Cd81","Rspo3")
all_plots = list()
for (mm in sel_markers) {
  p = FeaturePlot(scData.combined, features =mm, ncol = 1, order=T)  + NoAxes() + theme(legend.key.size = grid::unit(10, "points"))
  all_plots[[mm]] = p
}

cowplot::plot_grid(plotlist = all_plots, nrow=2, ncol=3, greedy = T)

pdf(file.path(result_folder, "Fig_2e.pdf"), width = 10, height = 7)
cowplot::plot_grid(plotlist = all_plots, nrow=2, ncol=3, greedy = T)
dev.off()

```


# Fig 3

## Fig 3a

```{r, fig.width=10, fig.height=5}
sel_markers = c("Grem1","Grem2","Mgp","Chrd","Fst","Nog",paste0("Bmp",1:7))
p = DotPlot(scData.combined, features = sel_markers, assay = "RNA", cols="RdYlBu") + theme(axis.text.x = element_text(angle=45, hjust=1))
print(p)

pdf(file.path(result_folder, "Fig_3a.pdf"), width = 10, height = 5)
print(p)
dev.off()

```

## Fig 3b 

```{r, fig.width=10, fig.height=7}
sel_markers =  c("Bmp2","Bmp5","Bmp7","Grem1","Grem2","Mgp")
all_plots = list()
for (mm in sel_markers) {
  p = FeaturePlot(scData.combined, features =mm, ncol = 1, order=T)  + NoAxes() + theme(legend.key.size = grid::unit(10, "points"))
  all_plots[[mm]] = p
}

cowplot::plot_grid(plotlist = all_plots, nrow=2, ncol=3, greedy = T)

pdf(file.path(result_folder, "Fig_3b.pdf"), width = 10, height = 7)
cowplot::plot_grid(plotlist = all_plots, nrow=2, ncol=3, greedy = T)
dev.off()

```


# Supp Fig. 3

## Supp Fig 3a


```{r, fig.width=14, fig.height=3}
sel_markers =   c("Pecam1","Col1a2","Pdgfra","Acta2","S100b","Elavl4")
all_plots = list()
for (mm in sel_markers) {
  p = FeaturePlot(scData.combined, features =mm, ncol = 1, order=T)  + NoAxes() + theme(legend.key.size = grid::unit(10, "points"))
  all_plots[[mm]] = p
}

cowplot::plot_grid(plotlist = all_plots, nrow=1, ncol=6, greedy = T)

pdf(file.path(result_folder, "Fig_S3a.pdf"), width = 14, height = 3)
cowplot::plot_grid(plotlist = all_plots, nrow=1, ncol=6, greedy = T)
dev.off()

```

## Supp Fig 3b

```{r, fig.width=10, fig.height=5}
sel_markers = c(paste0("Wnt",c("2","2b","4","5a","5b","6","9a","10a","11")), paste0("Rspo",1:3), "Dkk2","Dkk3","Sfrp1","Sfrp4","Wif1","Frzb")
p = DotPlot(scData.combined, features = sel_markers, assay = "RNA", cols="RdYlBu") + theme(axis.text.x = element_text(angle=45, hjust=1))
print(p)

pdf(file.path(result_folder, "Fig_S3b.pdf"), width = 10, height = 5)
print(p)
dev.off()

```


## Heatmap Fig X

Heatmap of in vivo data showing the following gene expression in telocytes (fibroblast 2) and trophocytes (fibroblast 1 and 3):

  - (Telocyte markers) Wnt4, Gata3, Smad6, Gpr20, Rab3b, Sox6, Bmp5, Bmp7
  - (Trophocyte markers) C3, Ill3, Sfrp1, Angptl7, Cfb, C4b, Serpina3a, Il6


```{r}
sel_genes = strsplit(c("Wnt4, Gata3, Smad6, Gpr20, Rab3b, Sox6, Bmp5, Bmp7, C3, Il33, Sfrp1, Angptl7, Cfb, C4b, Serpina3a, Il6"), ", ") |> unlist()

sel_mat = AverageExpression(scData.combined, assays = c("RNA"), features = sel_genes)

#pheatmap(sel_mat[["SCT"]], scale="row", cluster_rows = F, main="SCT")
pheatmap(sel_mat[["RNA"]], scale="row", cluster_rows = F, cluster_cols = F, main="Average expression (scaled)")

pheatmap(sel_mat[["RNA"]][, c("Fibroblast 1","Fibroblast 2", "Fibroblast 3")], scale="row", cluster_rows = F, cluster_cols = F, main="Average expression")




```


# Software versions

```{r}
sessionInfo()
```














