---
title: "GSE172261 Colon mesenchyme, mouse scData Preprocessing"
author: "Hilmar Berger"
output: 
 html_document:
    toc: true
    toc_depth: 3
    toc_float: false
    number_sections: true
    code_folding: hide
    
pdf_document:
    fig_caption: true
date: "`r format(Sys.time(), '%d %B, %Y %H:%M:%S')`"
---

```{r, message=FALSE}
library(Seurat)
library(ggplot2)

library(readxl)
library(knitr)
library(data.table)
library(cowplot)

library(homologene)

library(pheatmap)

#library(future)
#plan("multiprocess", workers = 2)

options("save.defaults" = list(compress=TRUE))

data_storage_folder = "../../../Results/Jasso_Stroma"
```

```{r}
load(file.path(data_storage_folder, "All_Filtered_scData.Rdata"))
all_ge_data_filtered$sample_label = all_ge_data_filtered$orig.ident
table(all_ge_data_filtered$orig.ident)
```

# Estimation and correction of cell cycle effect

We use a list of known cell cycle genes (provided by the Seurat package and translated to Mus musculus gene symbols) to estimate the contribution of cell cycle phase on the global gene expression.

<!-- ```{r} -->
<!-- # those are the mouse homologues of human cell cycle genes that come with Seurat  -->
<!-- cell_cycle_genes = read.table("../regev_lab_cell_cycle_genes_OK_working_with_phase.txt", sep="\t", header=F, stringsAsFactors = F) -->
<!-- colnames(cell_cycle_genes) = c("Symbol","Phase") -->
<!-- s.genes = subset(cell_cycle_genes, Phase=="S")$Symbol -->
<!-- g2m.genes = subset(cell_cycle_genes, Phase=="G2M")$Symbol -->
<!-- ``` -->

```{r}
scData = all_ge_data_filtered
scData <- NormalizeData(object = scData, normalization.method = "LogNormalize", scale.factor = 1e4)
gc()
scData <- FindVariableFeatures(object = scData, selection.method = "vst")
gc()
scData <- ScaleData(object = scData, features = rownames(scData))
gc()
```

## Association of major variance components with cell cycle

```{r}
scData <- RunPCA(scData, features = VariableFeatures(scData), ndims.print = c(1:20), nfeatures.print = 20)
#scData <- RunPCA(scData, features = VariableFeatures(scData), ndims.print = c(1:100), nfeatures.print = 20)

# check top 100 positively loaded genes for PC_3 for cell cycle association in MSigDB
#xx = scData@reductions$pca@feature.loadings
#cat(paste(names(tail(sort(xx[, "PC_3"]), 100)), collapse="\n"))
gc()
```


<!-- ```{r, fig.width=12} -->
<!-- DimHeatmap(scData, dims = c(1, 7, 13)) -->
<!-- ``` -->

## Cell cycle scoring

We use the Seura cell cycle scoring method to score association with S and G2/M phases and assign cells to G1, S or G2/M

```{r}
mouse.cc.genes = lapply(cc.genes.updated.2019, function(x) unique(human2mouse(x, db = homologeneData2)$mouseGene))
#scData <- CellCycleScoring(scData, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
scData <- CellCycleScoring(scData, s.features = mouse.cc.genes$s.genes, g2m.features = mouse.cc.genes$g2m.genes, set.ident = TRUE)

# view cell cycle scores and phase assignments
#head(scData[[]])
boxplot(scData$S.Score ~ scData$Phase)
boxplot(scData$G2M.Score ~ scData$Phase)
```

```{r, fig.height=6}
RidgePlot(scData, features = c("Pcna", "Top2a", "Mcm6", "Mki67"), ncol = 4)
```

```{r, fig.width=13, fig.height=8}
# PCA on cell cycle genes reveals
scData <- RunPCA(scData, features = unlist(mouse.cc.genes))
DimPlot(scData) + geom_hline(yintercept = 0, lty=2) + geom_vline(xintercept = 0, lty=2)
DimPlot(scData,dims = c(2,3) )
DimHeatmap(scData, dims = c(1, 2))
```


```{r}
m = table(scData$Phase, scData$sample_label)
kable(m)

tmp = as.matrix.data.frame(m)
colnames(tmp) = colnames(m)
rownames(tmp) = rownames(m)
mr = sweep(tmp, 2, apply(tmp,2,sum),"/")
kable(round(mr,2))

pheatmap(mr, scale="none", main="Relative contributions of cell cycle phases per experimental group", breaks=seq(0,1,by=0.01))
```

# Final preprocessing

Since there are several independent batches of samples with partially very different outcomes in terms of processing quality and cell viability we will use data set integration (Stuart and Butler et al., Cell 2019, doi: 10.1016/j.cell.2019.05.031) as implemented in Seurat v3.


```{r}
old_meta = scData@meta.data
scData = all_ge_data_filtered

# GSE172261 metadata define a batch, so let's use it here
#scData@meta.data$batch = scData$orig.ident
scData@meta.data$Phase = old_meta$Phase
scData@meta.data$S.Score = old_meta$S.Score
scData@meta.data$G2M.Score = old_meta$G2M.Score

sc.list <- SplitObject(scData, split.by = "batch")

sc.list <- lapply(X = sc.list, FUN = function(x) {
    x <- NormalizeData(x)
    x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})
```



```{r}
sc.anchors = FindIntegrationAnchors(object.list = sc.list, dims = 1:20)
gc()
scData.combined <- IntegrateData(anchorset = sc.anchors, dims = 1:20)
gc()
```


```{r}
DefaultAssay(scData.combined) <- "integrated"

scData.combined <- ScaleData(scData.combined, verbose = TRUE)
gc()
scData.combined <- RunPCA(scData.combined, npcs = 30, verbose = TRUE)
gc()
# t-SNE and Clustering
scData.combined <- RunUMAP(scData.combined, reduction = "pca", dims = 1:20)
scData.combined <- FindNeighbors(scData.combined, reduction = "pca", dims = 1:20)
scData.combined <- FindClusters(scData.combined, resolution = 0.1)
```

```{r}
save(scData.combined, file=file.path(data_storage_folder, "GSE172261_ColonMesenchyme_mouse.combined.RData"), compress = T)
```

# Global clusters

```{r, fig.width=14}
DimPlot(scData.combined, reduction = "umap", group.by = "condition")

p1 <- DimPlot(scData.combined, reduction = "umap", group.by = "condition")
p2 <- DimPlot(scData.combined, reduction = "umap", label = TRUE)
plot_grid(p1, p2)

DimPlot(scData.combined, reduction = "umap", split.by = "sample_label", label = T)
```

## Cluster assignments

```{r}
m = table(scData.combined@meta.data$seurat_clusters, scData.combined@meta.data$sample_label)
kable(m)
```


```{r, fig.height=10, fig.width=8}
mm = as.matrix.data.frame(m)
colnames(mm) = colnames(m)
rownames(mm) = rownames(m)
mm_rel = sweep(mm, 2, apply(mm, 2,sum), "/")
pheatmap(mm_rel, main="Relative single cell contribution of each cluster to condition totals", cluster_cols = F)
```


# Software versions

```{r}
sessionInfo()
```

